### PR TITLE
Use FMIS/BBR id to get commonscat from heritage database

### PR DESCRIPTION
**Pull request by lokal-profil**
Also:
* store bbr/fmis ids and re-factor process_depicted
* set FMIS default category on municipal level
* attempt municipal categorisation on tags

Task: [T164572](https://phabricator.wikimedia.org/T164572)
Task: [T164719](https://phabricator.wikimedia.org/T164719)

### CODE REVIEW

File: KMB/kmb_massload.py [hunk 1]
```diff
@@ -63,8 +63,9 @@ def parser(dom, A):
             A['longitude'] = coords[0][:8]
         else:
             A['problem'].append("Coord was not a point: '{0}'".format(cs))
-    # do visualizes separately need not be shm/fmi/bbr etc. can be multiple
-    A['bbr'] = A['fmis'] = False
+    # do visualizes separately need not be fmi/bbr etc. can be multiple
```
Comment (Vesihiisi) [inline KMB/kmb_massload.py]: Could you rephrase this? Not sure what it actually means.

File: KMB/kmb_massload.py [hunk 2]
```diff
@@ -99,49 +100,67 @@ def parser(dom, A):
...
-        A['bbr'] = True
-        crop = len('http://kulturarvsdata.se/raa/bbrb/')
-        A['avbildar'].append('{{BBR|%s|b}}' % url[crop:])
-    elif url.startswith('http://kulturarvsdata.se/raa/bbrm/'):
-        A['bbr'] = True
-        crop = len('http://kulturarvsdata.se/raa/bbrm/')
-        A['avbildar'].append('{{BBR|%s|m}}' % url[crop:])
-    else:
-        A['avbildar'].append(url)
+    mapping = {
```
Comment (Vesihiisi) [inline KMB/kmb_massload.py]: At some point, when a mapping grows past a certain size, it's more convenient to store it in a separate file -- I don't think it's happening here yet, but this might well be the treshold?

File: KMB/make_KMB_info.py [hunk 3]
```diff
@@ -476,13 +524,77 @@ def get_description(self):
...
+                self.content_cats.add('Listed buildings in Sweden')
+
+        if found_commonscat:
+            self.needs_place_cat = False
+
+        return found_commonscat
+
+    def make_default_fmis_category(self, cache):
+        """
+        Set FMIS default categories based municipality.
```
Comment (Vesihiisi) [inline KMB/make_KMB_info.py]: based _on_ municipality

File: KMB/make_KMB_info.py [hunk 4]
```diff
@@ -476,13 +524,77 @@ def get_description(self):
...
+
+        Makes a guess for the category name. If not found it defaults to
+        County (Län) + Land (Landskap) categories.
```
Comment (Vesihiisi) [inline KMB/make_KMB_info.py]: _landskap_ is _province_ on Commons.

File: KMB/make_KMB_info.py [hunk 5]
```diff
@@ -476,13 +524,77 @@ def get_description(self):
...
+
+        Populates self.content_cats and modifies self.needs_place_cat.
```
Comment (Vesihiisi) [inline KMB/make_KMB_info.py]: [Sorry.](http://imgur.com/gallery/tcbx9)

File: KMB/make_KMB_info.py [hunk 6]
```diff
@@ -476,13 +524,77 @@ def get_description(self):
...
+
+        :param cache: cache for category existence
+        """
+        test_cat = None
+        if self.kommunName:
+            test_cat = 'Archaeological monuments in {} Municipality'.format(
+                self.kommunName)
+            if not self.kmb_info.category_exists(test_cat, cache):
+                test_cat = 'Archaeological monuments in {}'.format(
+                    self.kommunName)
```
Comment (Vesihiisi) [inline KMB/make_KMB_info.py:595]: I wonder if this can ever end up badly. I.e. are there cases when [stuff in $place] would _not_ be a good substitute for [stuff in $place Municipality]? It would require there existing a non-municipal place that just happens to share a name with a municipal one, but I can't think of an example off the top of my head. Quite possibly not an actual problem in practice?

File: KMB/kmb_massload.py [hunk 7]
```diff
@@ -99,49 +100,67 @@ def parser(dom, A):
...
+            entry[rule['type_key']].add(idno)
+            entry['avbildar'].append(rule['template'] % idno)
+            found = True
+            break
+
+    if not found:
+        pattern = 'http://kulturarvsdata.se/raa/bbr/'
+        if url.startswith(pattern):
+            idno = url[len(pattern):].strip()
+            template = '{{BBR|%s%s}}'
```
Comment (Vesihiisi) [inline KMB/kmb_massload.py]: Can all these templates be stored as variables somewhere? Might be convenient to just go with with `bbr_template`, especially since we're doing some formatting on it. I.e. be able to write `bbr_template.type = 'm'` or sth like that.

File: KMB/make_KMB_info.py [hunk 8]
```diff
@@ -494,11 +606,25 @@ def make_tag_categories(self, cache):
...
                         tag_map[tag].get('SE'):
                     cat = tag_map[tag].get('SE')
+
+                    # attempt municipal categorisation
+                    if self.kommunName:
+                        test_cat = tag_map[tag].get('SE').replace(
+                            'Sweden', '{} Municipality'.format(
+                                self.kommunName))
+                        if self.kmb_info.category_exists(test_cat, cache):
+                            self.needs_place_cat = False
```
Comment (Vesihiisi) [inline KMB/make_KMB_info.py:644]: Ultimately, what happens to items with `self.needs_place_cat == True`?

File: KMB/kmb_massload.py (hunk 1 above)
Comment (lokal-profil) [reply KMB/kmb_massload.py]: I shortened it. It is mainly a leftover from before.

File: KMB/make_KMB_info.py (hunk 4 above)
Comment (lokal-profil) [reply KMB/make_KMB_info.py]: Done. True _Land_ is _Landsdel_

File: KMB/make_KMB_info.py (hunk 3 above)
Comment (lokal-profil) [reply KMB/make_KMB_info.py]: Done

File: KMB/kmb_massload.py (hunk 2 above)
Comment (lokal-profil) [reply KMB/kmb_massload.py]: See next comment

File: KMB/kmb_massload.py (hunk 7 above)
Comment (lokal-profil) [reply KMB/kmb_massload.py]: Implemented a BbrTemplate and FmisTemplate class and moved most of the logic there.

File: KMB/make_KMB_info.py (hunk 5 above)
Comment (lokal-profil) [reply KMB/make_KMB_info.py]: =)

File: KMB/make_KMB_info.py (hunk 6 above)
Comment (lokal-profil) [reply KMB/make_KMB_info.py:595]: In general there would be a risk of there e.g. being a place in another country with the same name. On the other hand that is equally true when you have the _Muncipality_ ending.

There worst that I think is likely to happen is something like the following (made up example though):
* Say the category for Göteborg Municipality is actually called _Archaeological monuments in Gothenburgh Municipality_
* We therefore don't find  _Archaeological monuments in Göteborg Municipality_
* We guess  _Archaeological monuments in Göteborg_ which is actually a subcategory of  _Archaeological monuments in Gothenburgh Municipality_ for the downtown areas.

In this case I can live with the miss categorisation since it is primarily the fault of unclear categories and the destination where we stick the images is close to correct.

File: KMB/make_KMB_info.py (hunk 8 above)
Comment (lokal-profil) [reply KMB/make_KMB_info.py:644]: On [L381](https://github.com/lokal-profil/upload_batches/pull/11/files/39a31a810c924e0bc3c0f8c35f2d4f6883d97274#diff-c27a0e60250eb4480e9cb3a67a168625R381) the plan is to introduce a mechanism which adds the municipal or parish category (if Sweden) and possibly a country category otherwise.

**Review by Vesihiisi — APPROVED** [review]
